File: 2022/2022W06/preppindataW06.R
library(readxl)
library(dplyr)
library(tidyr)
library(splitstackshape)
library(stringr)
library(fuzzyjoin)
options("scipen"=100, "digits"=14)

xlsx <- "C:/Data/PreppinData/7 letter words.xlsx"

scores <- read_excel(xlsx, 'Scrabble Scores') %>%
  separate(`Scrabble`, c('Points', 'TileFreqDtl'), sep='point(s*):', convert=TRUE) %>%
  cSplit(., 'TileFreqDtl', ',', type.convert=FALSE) %>%
  pivot_longer(cols = -'Points', names_to = 'ToDrop', values_to = 'TileFreq', values_drop_na = TRUE) %>%
  mutate('Tile' = str_extract(`TileFreq`,'(\\w+)(?=\\s)'),
         'Frequency' = as.integer(str_extract(`TileFreq`,'(\\d+)$')),
         'LN Chance' = log(`Frequency`/sum(`Frequency`)))

words <- read_excel(xlsx, '7 letter words') %>%  
  mutate('Temp' = toupper(str_replace_all(`7 letter word`,'\\B',','))) %>%
  cSplit(., 'Temp', ',', type.convert=FALSE) %>%
  pivot_longer(cols = -c('7 letter word'), names_to = 'ToDrop', values_to = 'Letter', values_drop_na = TRUE) %>%
  group_by(`7 letter word`, `Letter`) %>%
  mutate('Letter Count' = n()) 

final <- fuzzy_inner_join(words, scores, 
                          by = c('Letter'='Tile','Letter Count'='Frequency'), match_fun = list(`==`, `<=`)) %>%
  group_by(`7 letter word`) %>%
  filter(n()==7) %>%
  summarise('% Chance' = exp(sum(`LN Chance`)),
            '% Chance Temp' = round(`% Chance`,15),
            'Total Points' = sum(`Points`)) %>%
  mutate('Likelihood Rank' = dense_rank(desc(`% Chance Temp`)),
         'Points Rank' = dense_rank(desc(`Total Points`))) %>%
  select(c('Points Rank', 'Likelihood Rank', '7 letter word', '% Chance', 'Total Points'))
  
View(final)

File: 2022/2022W06/PreppindataW06.py
import pandas as pd
import re
import numpy as np

xlsx = pd.ExcelFile(r"C:\Data\PreppinData\7 letter words.xlsx")

scores = pd.read_excel(xlsx,'Scrabble Scores')
scores['Points'] = scores['Scrabble'].apply(lambda x: int(re.search('^(\d+)',x).group(1)))
scores['TileFreq'] = scores['Scrabble'].str.extract('(?<=:\s)(.*)')
scores = pd.concat([scores['Points'], 
                   pd.DataFrame([map(str.strip, x) for x in scores['TileFreq'].str.split(',').values.tolist()])], 
                   axis=1, sort=False)
scores = scores.melt(id_vars='Points', value_name='TileFreq', var_name='ToDrop').dropna().drop('ToDrop', axis=1)    
scores['Tile'] = scores['TileFreq'].apply(lambda x: re.search('^(\w+)',x).group(1))
scores['Frequency'] = scores['TileFreq'].apply(lambda x: float(re.search('(\d+)$',x).group(1)))
scores['LN Chance'] = np.log(scores['Frequency'] / scores['Frequency'].sum())

words = pd.read_excel(xlsx,'7 letter words') 
words = pd.concat([words, 
                   pd.DataFrame([map(str.upper, x) for x in words['7 letter word'].str.split('').values.tolist()])], 
                   axis=1, sort=False)
words = words.melt(id_vars='7 letter word', value_name='Letter', var_name='ToDrop').drop('ToDrop', axis=1).replace('', np.NaN).dropna()
words['Letter Count'] = words['Letter'].groupby([words['7 letter word'],words['Letter']]).transform('count')

final = pd.merge(words, scores, left_on='Letter', right_on='Tile')
final = final[final['Frequency']>=final['Letter Count']]
final['Match Letter'] = final['Letter'].groupby(final['7 letter word']).transform('count')
final = final[final['Match Letter']==7]

final = final.groupby('7 letter word', as_index=False).agg({'Points':'sum', 'LN Chance':'sum'}).rename(columns={'Points':'Total Points'})
final['% Chance'] = np.exp(final['LN Chance'])
final['% Chance Temp'] = round(final['% Chance'],15)
final['Likelihood Rank'] = final['% Chance Temp'].rank(method='dense',ascending=False).astype(int)
final['Points Rank'] = final['Total Points'].rank(method='dense',ascending=False).astype(int)
final = final[['Points Rank', 'Likelihood Rank', '7 letter word', '% Chance', 'Total Points']]
